Parse rustls.h to generate an API docs website

cpu opened the pull request:
### librustls: add missing API docs
A few elements of our public API had no associated docs. This is going to be flagged as an error by the API doc generation tooling, so let's fix it up-front with some short doc additions.

### add docgen tools binary

This commit adds a `docgen` binary to the tools crate. This tool can generate a .json file describing the API based on parsing the `cbindgen` generated `rustls.h` header file using a tree-sitter C grammar. 

The produced JSON can in turn be used to generate markdown web documentation, or any other format required.

Along with generating docs this tool applies some basic policy. In particular it requires that **all** public API items be documented or it will produce an error instead of the JSON data. We can extend this in the future to require (for example) describing arguments and return values in doc comments.

The generated JSON has the following structure:

```json
{
  "structs": [ ... ],
  "functions": [ ... ],
  "callbacks": [ ... ],
  "enums": [ ... ],
  "externs": [ ... ],
  "aliases": [ ... ]
}
```

Where each key is a general category of item:

* Structure definitions
* API function prototypes
* Callback typedefs
* Enum typedefs
* Extern typedefs
* Simple type alias typdefs

Within each category items are described like:

```json
{
  "anchor": "rustls-accepted",
  "comment": "A parsed ClientHello produced by a ...",
  "feature": null,
  "name": "rustls_accepted",
  "text": "```c\nstruct rustls_accepted\n```"
},
```

The anchor field is useful for creating anchor links that identify the item. The name is the actual name of the item. The comment is the C block comment content (with the block comment syntax removed, but other whitespace left as-is). If the item is surrounded by an ifdef guard for a crate feature, then the required feature name is used as the feature field. Lastly the text field holds the actual text that defines the item in rustls.h (without comment) as a C formatted markdown code block.

Backtick references in comments that match anchor names (after replacing
`_` with `-` and stripping an optional `()` suffix) are automatically
hyperlinked.

### add an API docs website powered by Zola
This commit adds a [Zola](https://www.getzola.org/) website that can template the API docs JSON
data into a nice website showing off our API.

The choice of Zola (and the initial CSS/config) are inspired by the https://rustls.dev/ website which also uses Zola. In the future we can add other pages (e.g. higher level tutorial/getting started content).

To use:
```
cargo run -p rustls-ffi-tools --bin docgen > website/static/api.json
cd website && zola serve
open http://locahost:1111
```

### ci: add docs page deployment workflow

Lifted from the equivalent config in the rustls repo, and customized for rustls-ffi.

Updates #215

cpu commented:
> ci: add docs page deployment workflow

Not sure the best way to demonstrate this working without merging first. Probably running `zola` locally is the best bet for now?

cpu commented:
> rustls-ffi / Build+Test (ubuntu-latest, clang, nightly, aws-lc-rs) (push) Failing after 3m

Odd. Out of time for today but I'll look into this soon.

cpu commented:
>> rustls-ffi / Build+Test (ubuntu-latest, clang, nightly, aws-lc-rs) (push) Failing after 3m
>
> Odd. Out of time for today but I'll look into this soon.

Not so odd. I broke the `ech_fetch` util invocation moving things around, and also needed to exclude ECH tests from the MSRV runs because the tools workspace w/ ech_fetch needs a higher MSRV for tree-sitter.

cpu commented:
Here's a preview I made with a test repo: https://cpu.github.io/rustls-ffi-docs/

jsha reviewed (commented):
Haven't reviewed the code in detail but the concept and the output look good to me 👍🏻

cpu commented:
> Haven't reviewed the code in detail but the concept and the output look good to me 

Thanks for taking a look, much appreciated :bow:

ctz commented on website/config.toml:
```diff
@@ -0,0 +1,19 @@
+# The URL the site will be built for
+# TODO(@cpu): set up https://ffi.rustls.dev DNS
+base_url = "https://rustls.github.io/rustls-ffi/"
```
FWIW I've set up the DNS for this

cpu replied:
Thank you! It seems to be showing as "DNS Check in Progress" in the [repo settings page](https://github.com/rustls/rustls-ffi/settings/pages). I'll check again in a bit.

cpu replied:
✅  DNS check successful

cpu replied:
Updated the site config to use `ffi.rustls.dev`.

ctz reviewed (commented):
sorry, these review comments were trapped in "finish your review" purgatory

cpu commented on .github/workflows/test.yaml:
```diff
@@ -130,7 +130,8 @@ jobs:
 
       - name: Run ECH connect test
-        if: matrix.crypto == 'aws-lc-rs' # No HPKE in ring
+        # No HPKE in ring, tools package w/ ech-fetch util needs higher MSRV.
+        if: matrix.crypto == 'aws-lc-rs' && matrix.rust != '1.73'
```
Done

cpu commented:
> rustls-ffi / Windows (aws-lc-rs, Debug) (push) Failing after 5m

Filed https://github.com/rustls/rustls-ffi/issues/514 for this. There's something flaky going on for this test on Windows runners.

cpu commented:
> cpu force-pushed the cpu-api-docs-clean branch

I added detection of `#if defined(DEFINE_XXX)` feature flag gates to document the associated items as requiring feature `xxx`. E.g. see [`rustls_connection_read_2`](https://cpu.github.io/rustls-ffi-docs/#rustls-connection-read-2) that requires `read_buf`, or [`rustls_ring_crypto_provider`](https://cpu.github.io/rustls-ffi-docs/#rustls-ring-crypto-provider) that requires `ring`.

I also implemented basic cross-linking of items in comments using [tree-sitter-md](https://crates.io/crates/tree-sitter-md) to parse the comments like we parse the `.h`. Comment text like `` `rustls_result` ``, or `` `rustls_crypto_provider_free()` `` are automatically hyper-linked to `#rustls-result` and `#rustls-crypto-provider-free`.

Both updates are reflected on my preview site: https://cpu.github.io/rustls-ffi-docs/

My plan now is to stop fiddling with this, land this branch after a +1 review, and then do a pass through the existing documentation comments to try and tidy them up using more backtick refs and so on.

cpu commented:
> cpu force-pushed the cpu-api-docs-clean branch 

rebased on main.
